Fix content-visibility-048.html

rwlbuis opened the pull request:
#### d72e06f60a0cc794bb5299e2c0394aa10e3ed662
<pre>
Fix content-visibility-048.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=261032">https://bugs.webkit.org/show_bug.cgi?id=261032</a>

Reviewed by Ryosuke Niwa.

Fix content-visibility-048.html by implementing [1], i.e. set the CSS target given the range.

The remaining failure in content-visibility-048.html is because the range resulting from the
fragment text directive is centered in the selection, but the same thing happens without
content-visibility.

[1] <a href="https://wicg.github.io/scroll-to-text-fragment/#invoking-text-directives">https://wicg.github.io/scroll-to-text-fragment/#invoking-text-directives</a> (Monkeypatching HTML § 7.4.6.3 Scrolling to a fragment)

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-target.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment.html: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-expected.txt: Added.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):

Canonical link: <a href="https://commits.webkit.org/267744@main">https://commits.webkit.org/267744@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/725eda64bca6b8d13d23b9520540ae92e6a56ca3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16441 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21186 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18049 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18548 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17785 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20218 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15327 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15996 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22609 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16336 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16165 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20461 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16746 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14183 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15868 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15837 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4183 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20234 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16587 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2326,6 +2326,11 @@ bool LocalFrameView::scrollToFragment(const URL& url)
+                while (commonAncestor && !is<Element>(commonAncestor))
+                    commonAncestor = commonAncestor->parentElement();
+                if (commonAncestor)
+                    document->setCSSTarget(downcast<Element>(commonAncestor));
```
I can't find where the spec says to set the `:target` pseudo class state on the common ancestor. Can you clarify where that is done?

rwlbuis replied:
Sure, see footnote [1], then the section "In scroll to the fragment, handle a indicated part that is a range....", there is step 3.6 which sets the target.
It is a bit hard to refer to this in the commit message since there are multiple mockeypatches for § 7.4.6.3.

nt1m replied:
Ah right, thanks

rniwa commented on Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
```diff
@@ -206,7 +209,7 @@ static std::optional<SimpleRange> rangeOfStringInRange(const String& query, Simp
         
         if (isNonSearchableSubtree(*currentNode)) {
-            if (auto newStart = NodeTraversal::nextSkippingChildren(*currentNode))
+            if (auto newStart = nextSkippingChildrenInComposedTreeIgnoringUserAgentShadow(*currentNode))
```
Composed tree means flat tree, it's different from [shadow-including tree order](https://dom.spec.whatwg.org/#concept-shadow-including-tree-order).

rwlbuis replied:
Thanks, that naming is always confusing to me. Anyway, I removed that part since it is not needed to fix the actual test I am trying to fix.

rniwa commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2326,6 +2326,11 @@ bool LocalFrameView::scrollToFragment(const URL& url)
                 auto range = highlightRanges.first();
+                auto* commonAncestor = commonInclusiveAncestor<ComposedTree>(range);
+                while (commonAncestor && !is<Element>(commonAncestor))
+                    commonAncestor = commonAncestor->parentElement();
```
Surely, we need to call this function at most once?

rwlbuis replied:
Makes sense, done.

rniwa commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2326,6 +2326,11 @@ bool LocalFrameView::scrollToFragment(const URL& url)
             
             if (highlightRanges.size()) {
                 auto range = highlightRanges.first();
+                auto* commonAncestor = commonInclusiveAncestor<ComposedTree>(range);
```
Please use RefPtr.

rwlbuis replied:
Done.

rwlbuis committed:
Fix content-visibility-048.html
https://bugs.webkit.org/show_bug.cgi?id=261032

Reviewed by Ryosuke Niwa.

Fix content-visibility-048.html by implementing [1], i.e. set the CSS target given the range.

The remaining failure in content-visibility-048.html is because the range resulting from the
fragment text directive is centered in the selection, but the same thing happens without
content-visibility.

[1] https://wicg.github.io/scroll-to-text-fragment/#invoking-text-directives (Monkeypatching HTML § 7.4.6.3 Scrolling to a fragment)

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-target.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment.html: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-expected.txt: Added.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):

Canonical link: https://commits.webkit.org/267744@main
